### PR TITLE
Add initial Windows implementation

### PR DESCRIPTION
**Pull request by Jake-Shadle**
- Add first pass at windows minidump writer
- Remove assertion info, it is functionally useless
- Always open process handle internally
- Use published crash-context

See https://github.com/EmbarkStudios/crash-handling/pull/9 where it is being used.

### CODE REVIEW

**Comment by Jake-Shadle** [comment]
Oops, looks like minidump has changed, I'll fix this tomorrow.

**Comment by Jake-Shadle** [comment]
Oh, and add windows tests as well.

**Comment by Jake-Shadle** [comment]
Hmm, so I go the in process dump test working easily, but the external one doesn't work due to some problems internally in MiniDumpWriteDump failing to do `ReadProcessMemory` for some reason. My assumption is that the child process is maybe exiting early, or at least some variable is falling out of scope somehow, but couldn't figure it out. I'm not super concerned with that as my own tests only use external process dumping and they have worked flawlessly https://github.com/EmbarkStudios/crash-handling/tree/main/minidumper-test

**Comment by marti4d** [comment]
Hi @Jake-Shadle. Aria asked me to help do a first-pass review of this, as she is currently very busy on another project. I should some review comments ready for you soon. Sorry about the wait!

**Comment by Jake-Shadle** [comment]
No worries, thanks!

File: src/windows/minidump_writer.rs
Comment (marti4d) [inline src/windows/minidump_writer.rs:43]: Is there a reason for preferring this instead of `PROCESS_ALL_ACCESS`?

It's interesting that this value works, as I don't see it mentioned in the [MSDN docs for OpenProcess](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-openprocess). `PROCESS_ALL_ACCESS` also has the advantage of being in `Win32_System_Threading`, so you wouldn't have to define the special constant above

File: src/windows/minidump_writer.rs
Comment (marti4d) [inline src/windows/minidump_writer.rs:70]: You could use [`GetCurrentProcess`](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-getcurrentprocess) instead. It never fails and always has `PROCESS_ALL_ACCESS`

File: src/windows/minidump_writer.rs
Comment (marti4d) [inline src/windows/minidump_writer.rs:107]: If this will interfere with in-process dumping, does in make sense to remove the `current_process()` function for now?

File: src/windows/minidump_writer.rs
Comment (marti4d) [inline src/windows/minidump_writer.rs:129]: nit - You could use gwrite for these and it will update the offset for you

File: src/windows/minidump_writer.rs
Comment (marti4d) [inline src/windows/minidump_writer.rs:142]: I have a suggestion that you can feel free to ignore: I think it would make the code a bit more clear to just have `fill_handle_stream()` return an `Option<Vec<u8>>` and move the creation of the `md::MINIDUMP_USER_STREAM` to the "if let" statement below, like so:

```
if let Some(buf) = &mut handle_stream {
    user_streams.push(md::MINIDUMP_USER_STREAM {
        Type: MINIDUMP_STREAM_TYPE::HandleOperationListStream as u32,
        BufferSize: buf.len() as u32,
        // Still not getting over the mut pointers here
        Buffer: buf.as_mut_ptr().cast(),
    });
}
```


That way it keeps the subfunction focused on creating the stream buffer and this function continues to focus on building up the user stream headers. It also more closely mirrors what you already did above with the `breakpad_info`

File: src/windows/minidump_writer.rs
Comment (marti4d) [inline src/windows/minidump_writer.rs:151]: Perhaps `user_stream_info` would be a bit more clear of a name? Seems a bit strange that `user_stream` contains `user_streams` and not the other way around

File: src/windows/minidump_writer.rs
Comment (marti4d) [inline src/windows/minidump_writer.rs:184]: typo 'hopfully'

File: src/windows/minidump_writer.rs
Comment (marti4d) [inline src/windows/minidump_writer.rs:230]: Isn't '0' a success code? Maybe I'm misunderstanding somehow, but this seems like it would run the "good path" when the call is unsuccessful?

File: src/windows/minidump_writer.rs
Comment (marti4d) [inline src/windows/minidump_writer.rs:246]: I don't think the first part is necessary - A single expression of `std::slice::from_raw_parts((v as *const T) as *const u8, std::mem::size_of::<T>())` seems a bit more idiomatic?

File: tests/windows_minidump_writer.rs
Comment (marti4d) [inline tests/windows_minidump_writer.rs:41]: It might be worth asserting that this succeeds, since if it doesn't it will invoke UB in the next line

**Review by marti4d — COMMENTED** [review]
This is very nice, and it's great that you also wrote tests for it! Thanks a bunch for this, and sorry about the long wait for this review.

Most of my comments below are just minor suggestions and nitpicks -- You can probably ignore them if you don't agree; however, I have 3 concerns:

1) I think there may be reversed logic in `windows/minidump_writer.rs#230`
2) You mention that `let mut user_streams = Vec::with_capacity(2);` may interfere with in-process dumping. If that's the case, does it make sense that we are providing a `current_process()` function to allow a behavior that won't work properly?
3) I saw your comments above, and for the record I also see the failure of the `dump_external_process` test with the same code. Does it make sense to disable that test until it's fixed?

WRT to the failing test, I've done a bit of investigation and it looks like if you pass `std::ptr::null()` in place of `exc_info` during the call to `MiniDumpWriteDump`, it works... So the issue seems to be stemming from that.

It's getting a bit late for me right now, but tomorrow I will continue investigating --It looks like it *should* work, so it's fairly intriguing that it doesn't.

File: src/windows/minidump_writer.rs
Comment (Jake-Shadle) [reply src/windows/minidump_writer.rs:43]: This was just the access that the original Breakpad code was using and it just worked so actually didn't even look at the documentation for this function, but can try changing it.

File: src/windows/minidump_writer.rs
Comment (Jake-Shadle) [reply src/windows/minidump_writer.rs:107]: This was just an offhand comment that might seem more ominous than it is, I don't think there's any reason to not have current process dumping, I'll just add a comment on it that users should note in-process dumping is less reliable just in general.

File: src/windows/minidump_writer.rs
Comment (Jake-Shadle) [reply src/windows/minidump_writer.rs:230]: Oops.

File: tests/windows_minidump_writer.rs
Comment (Jake-Shadle) [reply tests/windows_minidump_writer.rs:41]: `RtlCaptureContext` doesn't have a return value unfortunately.

File: src/windows/minidump_writer.rs
Comment (Jake-Shadle) [reply src/windows/minidump_writer.rs]: Yup seems to work just fine.

File: src/bin/test.rs
Comment (marti4d) [inline src/bin/test.rs:322]: nit: There also exists a [`GetCurrentThread()`](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-getcurrentthread) Win32 API that never fails and returns a handle that doesn't need to be closed and has `THREAD_ALL_ACCESS` permissions

**Review by marti4d — COMMENTED** [review]
Thanks for addressing the feedback so quickly! I have a few more optional suggestions, but as far as I'm concerned this code is ready to check in once we figure out why that test is broken.

I'm investigating that right now, so hopefully we can get that resolved quickly and merge this change 😀

**Comment by marti4d** [comment]
@Jake-Shadle I left a PR to fix the broken test on your Embark Studios fork. Once you merge that in, I think this PR should be good-to-go! 😀

File: src/bin/test.rs
Comment (marti4d) [inline src/bin/test.rs:325]: nit: You don't have to use `CloseHandle` on this because `GetCurrentThread()` returns a pseudohandle and MSDN guarantees that `CloseHandle` is a no-op.

So you could replace all 3 lines with this one-liner:

`GetThreadContext(GetCurrentThread(), &mut exception_context);`

**Review by marti4d — APPROVED** [review]
@Gankra Hey Aria! IMO this change is good-to-go and I'm confident to merge it in at anytime.

Thanks a lot, @Jake-Shadle. This is great :)

**Pull request merged**
